src/hivpy/circumcision.py

```diff
@@ -0,0 +1,182 @@
+        """
+        Initialise circumcision at birth for all born males.
+
+        This is an alternative birth circumcision initialisation
+        method to `init_birth_circumcision_all` and requires
+        the use of `update_birth_circumcision` at every time step
+        to work as expected.
+        """
+        male_born_population = population.index[(population[col.SEX] == SexType.Male)
+                                                & (population[col.AGE] >= 0.25)]
+        r = rng.uniform(size=len(male_born_population))
+        circumcision = r < self.prob_birth_circ
+        population.loc[male_born_population, col.CIRCUMCISED] = circumcision
+        # all circumcised males get a circumcision date of the start of the simulation
+        population.loc[population[col.CIRCUMCISED], col.CIRCUMCISION_DATE] = date
+
+    def update_birth_circumcision(self, population, time_step, date):
+        """
+        Update birth circumcision for newly born males.
+        COVID disruption is factored in.
+
+        Requires birth circumcision to be initialised with
+        `init_birth_circumcision_born` instead of
+        `init_birth_circumcision_all` in order to work as expected.
+        """
+        # covid disruption causes circumcision probability to be 0
+        if (self.covid_disrup_affected != 1) & (self.vmmc_disrup_covid != 1):
+            # assumes ages have already been incremented
+            newborn_males = population.index[(population[col.SEX] == SexType.Male)
+                                             & (population[col.AGE] >= 0.25)
+                                             & (population[col.AGE] - time_step.days / 365 < 0.25)]
+            r = rng.uniform(size=len(newborn_males))
+            circumcision = r < self.prob_birth_circ
+            population.loc[newborn_males, col.CIRCUMCISED] = circumcision
+            # newly circumcised males get the current date set as their circumcision date
+            circ_newborn_males = population.index[population[col.CIRCUMCISED]
+                                                  & population[col.CIRCUMCISION_DATE].isnull()]
+            population.loc[circ_newborn_males, col.CIRCUMCISION_DATE] = date
+
+    def update_vmmc(self, pop):
+        """
+        Update voluntary medical male circumcision intervention.
+        COVID disruption is factored in.
+        """
+        # keep date in circumcision for now for the sake of
+        # the function passed to transform_group
+        self.date = pop.date  # should year be float instead of int?
+
+        # no further circumcision
+        if (self.vmmc_disrup_covid == 1) \
+           | ((self.year_interv <= self.date.year)
+              & (self.circ_inc_rate_year_i == 2)) \
+           | ((self.year_interv + 5 <= self.date.year)
+              & (self.circ_inc_rate_year_i == 4)):
+            self.test_link_circ_prob = 0
+
+        # only begin VMMC after a specific year
+        elif self.mc_int < self.date.year:
```

Comment:
SAS code treats the calendar date as a float so 2008.25 for example is march 2008. So we should activate VMMC the beginning of `mc_int`, whereas this will wait for the start of the next year